### analysis the telemetry and timing info

build detection rules for pitstop and crash events


# Imports

In [1]:
%matplotlib inline

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import os
#os.environ["CUDA_VISIBLE_DEVICES"]="7"


## Load Data

In [2]:
import os
os.getcwd()


'/scratch/hpda/indycar/predictor/notebook/5.DeployModel'

indy2018-pitstops is the dataset of \$C command with two more columns

1. timestamp, the latest timestamp before this $C command is recieved
2. lapdist, the latest lap_distance record from the telemetry of this particular car


In [3]:
# this file ware created from the log
# <latest_timestamp> <latest_lapdistance> $C 
pitstops = pd.read_csv('indy2018-pitstops.csv')
data = pitstops.drop_duplicates(subset=['car_number', 'completed_laps'], keep='first')
pitstops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1398 entries, 0 to 1397
Data columns (total 23 columns):
timestamp               1398 non-null object
lapdist                 1398 non-null float64
rank                    1398 non-null int64
car_number              1398 non-null int64
unique_id               1398 non-null object
completed_laps          1398 non-null int64
elapsed_time            1398 non-null float64
last_laptime            1398 non-null float64
lap_status              1398 non-null object
best_laptime            1398 non-null float64
best_lap                1398 non-null object
time_behind_leader      1398 non-null float64
laps_behind_leade       1398 non-null object
time_behind_prec        1398 non-null float64
laps_behind_prec        1398 non-null object
overall_rank            1398 non-null object
overall_best_laptime    1398 non-null float64
current_status          1398 non-null object
track_status            1398 non-null object
pit_stop_count          1398 non-n

In [4]:
# to find 'strage' pitstops when they pit not near the pitlane entry
#data[(data['lapdist'] > 500) and (data['lapdist'] < 4000)]
data[(data['lapdist'] > 300)][(data['lapdist'] < 4000)]

/scratch/hpda/anaconda3/envs/predictor/lib/python3.7/site-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


timestamp  lapdist  rank  car_number unique_id  completed_laps  \
188   18:22:35.885  2964.75    26          17        20             140   
191   18:26:20.326  2964.75    26          17        20             142   
449   19:16:53.390  3818.00    22          21         5             191   
608   16:45:42.847   330.16    32          26        18              31   
611   17:06:56.916   334.03    28          26        18              53   
617   17:08:42.914   332.73    30          26        18              54   
620   17:26:57.902   333.00    27          26        18              70   
623   17:51:54.808   330.75    24          26        18             101   
628   18:15:43.236   332.06    26          26        18             134   
632   18:37:27.516   331.44    25          26        18             150   
635   18:45:33.045   334.81    24          26        18             156   
639   19:05:32.485   339.13    23          26        18             180   
644   19:16:39.206   327.56    23          26        18             190   
770   17:50:15.063  1629.47    28          32        10              98   
1213  19:04:43.245  2352.81    11           6        11             180   
1294  16:45:34.638   332.39     7          98         B              32   
1311  17:00:54.728   328.61    10          98         B              50   
1316  17:44:49.015   331.88    15          98         B              92   
1329  18:08:23.243   323.94    10          98         B             125   
1342  18:57:28.487   328.86    14          98         B             170   

      elapsed_time  last_laptime lap_status  best_laptime  ...  \
188      7170.2231      120.3700          P       41.0610  ...   
191      7394.4007      123.6573          P       41.0610  ...   
449     10427.4875      120.6266          P       40.8170  ...   
608      1356.9691       69.0226          P       41.9395  ...   
611      2631.0479      127.2128          P       41.0509  ...   
617      2736.8594      105.8115          P       41.0509  ...   
620      3831.9342      123.3099          P       41.0509  ...   
623      5328.9025       73.1547          P       41.0509  ...   
628      6757.2836       79.3177          P       40.8811  ...   
632      8061.7088      105.8029          P       40.8811  ...   
635      8547.2184      112.5915          P       40.8811  ...   
639      9746.4006       68.1319          P       40.8811  ...   
644     10413.4707      119.0397          P       40.5188  ...   
770      5229.1082       68.4199          P       41.5829  ...   
1213     9697.4589       63.0600          P       40.6059  ...   
1294     1348.5699       55.3328          P       41.0183  ...   
1311     2268.6474      111.4303          P       41.0183  ...   
1316     4902.9639       54.4188          P       41.0183  ...   
1329     6317.4104       54.6062          P       40.7202  ...   
1342     9262.5522       53.5500          P       40.7202  ...   

     time_behind_prec  laps_behind_prec overall_rank  overall_best_laptime  \
188           12.0121                 0           1A               41.0610   
191           28.3470                 0           1A               41.0610   
449            2.9782                 0           16               40.8170   
608            1.8259                 0           20               41.9395   
611            0.3681                 0           1C               41.0509   
617            8.6727                 0           1E               41.0509   
620            3.4335                 0           1B               41.0509   
623            3.7780                 0           18               41.0509   
628            3.1213                 0           1A               40.8811   
632           18.0769                 0           19               40.8811   
635           12.3147                 0           18               40.8811   
639            7.1977                 0           17               40.8811   
644          106.6098           

### Anomaly of pitstops

1. car 17. pitstop entry is weired at '2964.75', the root cause is that the telemetry lost from 18:21:57.732 to 18:26:32.928

2. car 26. pitstio entry is as large as '330.16' at lap 31, not near the pit lane entry. By investigating the telemetry, it seems that the lap_distance value for this car is not well calibrated. The maximum value of lap_distance of previous lap, lap 30, is only '3734' 


In [5]:

pitstops[pitstops['car_number']==17]

timestamp  lapdist  rank  car_number unique_id  completed_laps  \
166  16:47:58.640   249.34    28          17        20              34   
167  16:48:29.828  1905.80    28          17        20              34   
168  16:48:36.075  2467.78    28          17        20              34   
169  16:48:44.052  3238.73    28          17        20              34   
170  16:48:46.831  3518.17    31          17        20              34   
171  16:48:50.806  3888.69    32          17        20              34   
172  17:07:02.303   255.20    30          17        20              53   
173  17:08:06.842  3175.95    30          17        20              53   
174  17:08:31.731  4214.02    30          17        20              53   
175  17:26:58.632   256.34    28          17        20              70   
176  17:28:14.145  3459.63    28          17        20              70   
177  17:28:35.900  4295.44    28          17        20              70   
178  17:53:21.908   248.56    23          17        20             103   
179  17:53:55.808  2130.28    23          17        20             103   
180  17:53:56.897  2222.91    23          17        20             103   
181  17:54:06.060  3096.84    24          17        20             103   
182  17:54:11.900  3654.53    25          17        20             103   
183  17:54:13.900  3848.41    26          17        20             103   
184  18:17:13.713   246.50    26          17        20             136   
185  18:17:39.886  1472.75    26          17        20             136   
186  18:17:50.697  2437.50    26          17        20             136   
187  18:18:08.794  3566.31    27          17        20             136   
188  18:22:35.885  2964.75    26          17        20             140   
189  18:23:58.887  2964.75    26          17        20             140   
190  18:24:09.796  2964.75    26          17        20             140   
191  18:26:20.326  2964.75    26          17        20             142   
192  18:27:23.790  3530.33    26          17        20             142   
193  18:27:33.891  3958.29    26          17        20             142   
194  19:10:20.200   252.27    21          17        20             187   
195  19:10:42.009  1097.22    21          17        20             187   
196  19:10:58.012  1770.47    21          17        20             187   
197  19:11:35.016  3698.08    22          17        20             187   
198  19:16:46.854  4087.48    20          17        20             191   
199  19:16:52.119  4227.64    20          17        20             191   
200  19:18:01.017  3032.25    20          17        20             191   
201  19:18:07.849  3627.53    21          17        20             191   

     elapsed_time  last_laptime lap_status  best_laptime  ...  \
166     1492.8432       76.4384          P       42.0016  ...   
167     1492.8432       76.4384          P       42.0016  ...   
168     1492.8432       76.4384          P       42.0016  ...   
169     1492.8432       76.4384          P       42.0016  ...   
170     1492.8432       76.4384          P       42.0016  ...   
171     1492.8432       76.4384          P       42.0016  ...   
172     2636.1781      127.4500          P       41.0610  ...   
173     2636.1781      127.4500          P       41.0610  ...   
174     2636.1781      127.4500          P       41.0610  ...   
175     3832.4968      125.1808          P       41.0610  ...   
176     3832.4968      125.1808          P       41.0610  ...   
177     3832.4968      125.1808          P       41.0610  ...   
178     5416.0141       70.7782          P       41.0610  ...   
179     5416.0141       70.7782          P       41.0610  ...   
180     5416.0141       70.7782          P       41.0610  ...   
181     5416.0141       70.7782          P       41.0610  ...   
182     5416.0141       70.7782          P       41.0610  ...   
183     5416.0141       70.7782          P       41.0610  ...   
184     6847.9776       69.4364   